### PR TITLE
Support Deepspeed finetuning

### PR DESCRIPTION
**Pull request by minmingzhu**
TODO:
        Needs to implement various Deepspeed config. Maybe refer to ACCELERATE_DEEPSPEED_CONFIG_FILE parameter.

### CODE REVIEW

File: finetune/finetune.py [hunk 1]
```diff
@@ -14,17 +14,29 @@
...
+from torch.distributed.fsdp.fully_sharded_data_parallel import FullOptimStateDictConfig, FullStateDictConfig
 
 import sys
 
 sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
 import common
-from finetune.finetune_config import FinetuneConfig
+from finetune_config import FinetuneConfig
+
+DEEPSPEED_CONFIG = {
```
Comment (harborn) [inline finetune/finetune.py:26]: can we implement a function to generate deepspeed config?

File: finetune/finetune.py [hunk 2]
```diff
@@ -77,14 +95,18 @@ def train_func(config: Dict[str, Any]):
...
+                                             deepspeed_plugin=deepspeed_plugin)
     else:
         fsdp_plugin = None
-    accelerator = accelerate.Accelerator(
-        gradient_accumulation_steps=gradient_accumulation_steps, fsdp_plugin=fsdp_plugin
-    )
-    common.logger.info(
-        f"accelerator generate finish, accelerator device type = {accelerator.device}"
-    )
+        accelerator = accelerate.Accelerator(gradient_accumulation_steps=gradient_accumulation_steps,
```
Comment (harborn) [inline finetune/finetune.py:106]: no need fsdp_plugin here, just:
`accelerator = accelerate.Accelerator(gradient_accumulation_steps=gradient_accumulation_steps)`

File: finetune/finetune_config.py [hunk 3]
```diff
@@ -57,14 +57,14 @@ class Training(BaseModel):
 
     @validator("device")
     def check_device(cls, v: str):
-        devices = ["CPU", "GPU"]
+        devices = ["CPU", "GPU", "XPU"]
```
Comment (harborn) [inline finetune/finetune_config.py:60]: why add `XPU` here?

File: pyproject.toml [hunk 4]
```diff
@@ -53,6 +53,7 @@ gpu = [
     "intel_extension_for_pytorch==2.1.10+xpu",
     "oneccl_bind_pt==2.1.100+xpu",
     "dpctl==0.15.0"
```
Comment (harborn) [inline pyproject.toml:55]: miss a ','

File: common/trainer/default_trainer.py [hunk 5]
```diff
@@ -137,14 +139,25 @@ def prepare(self, model, tokenizer, dataset, optimizer, accelerator):
...
-            self.optimizer,
-            self.train_dataloader,
-            self.eval_dataloader,
-            self.lr_scheduler,
-        ) = accelerator.prepare(optimizer, train_dataloader, eval_dataloader, lr_scheduler)
+        # self.model = accelerator.prepare(model)
+
+        accelerate_mode = self.config.get("accelerate_mode")
+        if accelerate_mode in ["GPU_DEEPSPEED"] :
+            dummy_optimizer = DummyOptim(params=model.parameters(), lr=5e-5, weight_decay=0.0)
```
[the remark this answers is not in the document]
Comment (harborn) [reply common/trainer/default_trainer.py:144]: please check lr and weight_decay in yaml

File: common/trainer/default_trainer.py [hunk 6]
```diff
@@ -135,14 +137,25 @@ def prepare(self, model, tokenizer, dataset, optimizer, accelerator):
...
+
+            def _lr_scheduler_callable(optimizer):
+                return get_scheduler(
+                    name="linear",
```
Comment (harborn) [inline common/trainer/default_trainer.py:148]: scheduler name is configurable?
